boards/shared/Src/Io/Io_SharedAdc.c

```diff
@@ -1,169 +1,140 @@
- */
-static uint32_t vrefint_index = NUM_ADC_CHANNELS;
-
-/** @brief Keep track of whether the ADC has been initialized yet */
-static bool adc_initialized = false;
-
-/**
- * @brief  Configure the maximum ADC value based on ADC resolution
- * @param  hadc ADC handle
- * @return ErrorStatus
- */
-static ErrorStatus InitializeAdcMaxValue(ADC_HandleTypeDef *hadc);
+struct AdcInput
+{
+    ADC_HandleTypeDef *hadc;
+    size_t             vrefint_index;
+    uint32_t           num_active_channels;
+    uint16_t           adc_max_value;
+    uint16_t *         adc_readings;
+};
 
-/**
- * @brief  Return the VREFINT calibration value acquired by the ADC during the
- *         manufacturing process at VDDA = 3.3V (Note that this is a 16-bit
- *         value)
- * @return VDDA calibration data
- */
-static uint16_t GetVrefintCalibrationValue(void);
+// Pointer to the VREFINT measured by the ADC
+static uint16_t *measured_vrefint_address;
 
 /**
- * @brief Check if the ADC module has been initialized using
- *        Io_SharedAdc_StartAdcInDmaMode()
- * @note  This should be called at the beginning of functions that require ADC
- *        to be initialized
+ * Gets the ADC resolution for the given ADC handle
+ * @param hadc The ADC handle for the ADC we are initializing
+ * @return adc_max_resolution The maximum ADC resolution for the given ADC
+ * handle
  */
-static void CheckAdcInitialized(void);
+static ErrorStatus
+    Io_GetAdcResolution(ADC_HandleTypeDef *hadc, uint16_t *adc_max_resolution);
 
-static ErrorStatus InitializeAdcMaxValue(ADC_HandleTypeDef *hadc)
+static ErrorStatus
+    Io_GetAdcResolution(ADC_HandleTypeDef *hadc, uint16_t *adc_max_resolution)
 {
     switch (hadc->Init.Resolution)
     {
         case ADC_RESOLUTION_6B:
-            adc_max_value = MAX_6_BITS_VALUE;
+            *adc_max_resolution = MAX_6_BITS_VALUE;
             break;
         case ADC_RESOLUTION_8B:
-            adc_max_value = MAX_8_BITS_VALUE;
+            *adc_max_resolution = MAX_8_BITS_VALUE;
             break;
         case ADC_RESOLUTION_10B:
-            adc_max_value = MAX_10_BITS_VALUE;
+            *adc_max_resolution = MAX_10_BITS_VALUE;
             break;
         case ADC_RESOLUTION_12B:
-            adc_max_value = MAX_12_BITS_VALUE;
+            *adc_max_resolution = MAX_12_BITS_VALUE;
             break;
         default:
             return ERROR;
     }
+
     return SUCCESS;
 }
 
-static uint16_t GetVrefintCalibrationValue(void)
+struct AdcInput *
+    Io_SharedAdc_Init(ADC_HandleTypeDef *hadc, size_t vrefint_index)
 {
-    return (uint16_t)(*VREFINT_CALIBRATION_ADDRESS);
-}
+    assert(hadc != NULL);
 
-static void CheckAdcInitialized(void)
-{
-    if (adc_initialized != true)
-    {
-        Error_Handler();
-    }
-}
+    static size_t          alloc_index = 0;
+    static struct AdcInput adc_inputs[MAX_NUM_OF_ADC];
+    static uint32_t adc_readings[MAX_NUM_OF_ADC][MAX_NUM_OF_ADC_CHANNELS];
```

Comment:
I don't think this library should be allocating the buffer to which the DMA will be transferring data. For two reasons,

1. One of the main advantages of DMA is that we are moving data directly into where we need to. If we are statically allocating buffer and then exposing a getter function, then we probably should've have just used interrupt-based ADC. What we want to with ADC in DMA mode is something like:

```
// Io_CurrentSense.c


uint16_t adc_readings_in_volts[2];

// main.c
extern uint16_t adc_readings_in_volts[2];

int main() {
  Io_SharedAdc_Init(hadc2, VREFINT_RANK, adc_readings_in_volts);
  ..
} 

#define CURRENT_SENSE_CONVERSION_RATIO 1500
1kHzTask() {
  // Convert ADC readings to Amps
  for (int i = 0; i < 2; i++) {
  printf("The current at channel %u is %u\n", i, adc_readings_in_volts[i] * CURRENT_SENSE_CONVERSION_RATIO);  
  }
}


 
2.